sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java

```diff
@@ -292,8 +298,8 @@ public void evaluate() throws Throwable {
   public PipelineResult run() {
     checkState(
         enforcement.isPresent(),
-        "Attempted to run a pipeline while it's enforcement level was not set. Are you "
-            + "using TestPipeline without a @Rule annotation?");
+        "Is your TestPipeline declaration missing a @Rule annotation? Usage: "
```

Comment:
I think there's been enough pain that we don't want to require that this is used as a rule.

In general it's reasonable to inline a `TestPipeline` just as a `Pipeline` with some reasonable defaults; and when `run` is being called _most_ of the time we'll be in a good spot. Within the SDK & runners we probably want to use the `@Rule`, but don't generally want to require all users do so.

Reply:
Can you give a snippet of your test file?
What runner were you using to test your code?

The issue is likely elsewhere and that somewhere the automatic detection is failing. From the javadoc:
`Abandoned node detection is automatically enabled when a real pipeline runner (i.e. not a {@link CrashingRunner}) and/or a {@link NeedsRunner} or a {@link RunnableOnService} annotation are detected.`
If you step through the code with a debugger, what condition prevents the `enforcement` from being enabled?

Reply:
`    @Rule
    public final transient TestPipeline pipeline = TestPipeline.create();
    @Test
    public void decodeMessageFnBadJson() {
        String badJson = "bad json is bad[]{";
        PCollection<String> collection = pipeline.apply(Create.of(badJson))
                .apply(ParDo.of(new CollectionPipeline.DecodeMessageFn()).withOutputTags(messageTag, sideTags))
                .get(deadLetterTag);

        PAssert.that(collection).containsInAnyOrder(badJson);
        pipeline.run();
    }`
I stepped through and found that it was looking for some kind of options.
Based on the documentation you mention, I probably need to specify some kind of runner or create some sort of options class. In any case I think the original error message was more descriptive of the problem.

Reply:
It seems like based upon this logic: https://github.com/apache/beam/blob/b4d201820bdf911ae4ce091ce8c5d99a5c0aaab2/sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java#L279

You should either do one of these three things:
1) Annotate your method with `@Category(NeedsRunner.class)` and use JUnit categories to filter your tests to tests that need a runner and ones that don't. `NeedsRunner` is a special category that is used in Apache Beam integration testing to label tests that need a runner like direct/flink/spark so you could use the same mechanism there when running your tests.
2) Make sure that you set the --runner option to anything but the CrashingRunner. You could also set this programmatically via options.setRunner(DirectRunner.class) but better to use the environment variable beamTestPipelineOptions as described in TestPipeline.java
3) Explicitly call enableAbandonedNodeEnforcement on TestPipeline which is a manual override.

Feel free to open a PR that updates the messaging to make it comprehensible.